File: service/fetch.rb
require "mongo"
require "csv"

def getUserinfo (userinfo) 
  if (!userinfo) then 
    userinfo = {}
  end
  return {
    submit: userinfo["submit"],
    fname: userinfo["fname"],
    lname: userinfo["lname"],
    nname: userinfo["nname"],
    nationality: userinfo["nationality"],
    natId: userinfo["natId"],
    birthday: userinfo["birthday"],
    sex: userinfo["sex"],
    gender: userinfo["gender"],
    email: userinfo["email"],
    phone: userinfo["phone"],
    telegram: userinfo["phone"],
    line: userinfo["line"],
    whatapps: userinfo["whatapps"],
    instagram: userinfo["instagram"],
    facebook: userinfo["facebook"],
    emergencyName: userinfo["emergencyName"],
    emergencyPhone: userinfo["emergencyPhone"],
    medCond: userinfo["medCond"],
    medRequire: userinfo["medRequire"],
    allergy: userinfo["allergy"],
    dietary: userinfo["dietary"],
    seasick: userinfo["seasick"],
    carsick: userinfo["carsick"],
    religion: userinfo["religion"],
    other: userinfo["other"],
  }
end 

client = Mongo::Client.new(ENV["MONGOURI"], :database => "test")
$users = client[:users]
$teams = client[:teams]

def teamIndex (id)
  $teams.find.map{|team| team["_id"]}.find_index(id)
end

result = $users.find
  .map{ |user| {
    role: user["as"],
    userid: user["email"],

    observer_submit: user["observer"] ? user["observer"]["submit"] : nil,
    observer_org: user["observer"] ? user["observer"]["organization"] : nil,
    observer_checkin: user["observer"] ? user["observer"]["checkin"] : nil, 
    observer_exc1: user["observer"] ? user["observer"]["excursion1"] : nil,
    observer_exc2: user["observer"] ? user["observer"]["excursion2"] : nil,
    observer_exc3: user["observer"] ? user["observer"]["excursion3"] : nil,
    observer_exc4: user["observer"] ? user["observer"]["excursion4"] : nil,

    teams: user["teams"] ? user["teams"].map{|team| teamIndex(team)} : nil,
  }.merge(getUserinfo(user["userinfo"]))}

headers = result[0].keys
path = 'users.csv'
CSV.open(path, 'w', headers: headers, write_headers: true) do |csv|
  result.each { |row| csv << row.values }
end


result = $teams.find
  .map{ |team| {
    id: teamIndex(team["_id"]),
    team_submit: team["info"] ? team["info"]["submit"] : nil,

    team_school: team["info"] ? team["info"]["school"] : nil,
    team_address: team["info"] ? team["info"]["address"] : nil,
    team_contactname: team["info"] ? team["info"]["contactname"] : nil,
    team_contactemail: team["info"] ? team["info"]["contactemail"] : nil,
    team_checkin: team["info"] ? team["info"]["checkin"] : nil,
    team_room: team["info"] ? team["info"]["room"] : nil,
    team_excursion1: team["info"] ? team["info"]["excursion1"] : nil,
    team_excursion2: team["info"] ? team["info"]["excursion2"] : nil,
    team_excursion3: team["info"] ? team["info"]["excursion3"] : nil,
    team_excursion4: team["info"] ? team["info"]["excursion4"] : nil 
  }
    .merge(getUserinfo(team["contestant1"]).transform_keys{|key| ('c1_' + key.to_s).to_sym })
    .merge(getUserinfo(team["contestant2"]).transform_keys{|key| ('c2_' + key.to_s).to_sym })
    .merge(getUserinfo(team["contestant3"]).transform_keys{|key| ('c3_' + key.to_s).to_sym })
  }

headers = result[0].keys
path = 'teams.csv'
CSV.open(path, 'w', headers: headers, write_headers: true) do |csv|
  result.each {|row| csv << row.values }
end

File: service/fetch.py
import os
from dotenv import load_dotenv
import pymongo
import functools
import csv

foldl = lambda func, acc, xs: functools.reduce(func, xs, acc)

def optionalGet(dict, args) : 
  try: 
    return foldl (lambda dict, arg: dict.get(arg), dict, args)
  except AttributeError: 
    return None

# def teamIndex(teamid) : 
#     db = client["test"]
#     collection = db["teams"]
#     documents = collection.find()
#     print([x for x in documents])

def getUserInfo(document): 
    opGet = lambda args: optionalGet(document, args)
    result = {
        "submit": opGet(["submit"]),
        "fname": opGet(["fname"]),
        "lname": opGet(["lname"]),
        "nname": opGet(["nname"]),
        "nationality": opGet(["nationality"]),
        "natId": opGet(["natId"]),
        "birthday": opGet(["birthday"]),
        "sex": opGet(["sex"]),
        "gender": opGet(["gender"]),
        "email": opGet(["email"]),
        "phone": opGet(["phone"]),
        "telegram": opGet(["phone"]),
        "line": opGet(["line"]),
        "whatapps": opGet(["whatapps"]),
        "instagram": opGet(["instagram"]),
        "facebook": opGet(["facebook"]),
        "emergencyName": opGet(["emergencyName"]),
        "emergencyPhone": opGet(["emergencyPhone"]),
        "medCond": opGet(["medCond"]),
        "medRequire": opGet(["medRequire"]),
        "allergy": opGet(["allergy"]),
        "dietary": opGet(["dietary"]),
        "seasick": opGet(["seasick"]),
        "carsick": opGet(["carsick"]),
        "religion": opGet(["religion"]),
        "rel_cere": opGet(["relCeremony"]),
        "other": opGet(["other"]),
    }
    return result

def getUser(documents):
    i = 0
    result = []
    for document in documents : 
        opGet = lambda args: optionalGet(document, args) 
        if i%100 == 0: print(i, 'rows done')
        i += 1

        result.append({ 
            "role": opGet(["as"]), 
            "userid": opGet(["email"]),
            "observer_submit": opGet(["observer", "submit"]),
            "observer_org": opGet(["observer", "organization"]), 
            "observer_checkin": opGet(["observer", "checkin"]),  
            "observer_exc1": opGet(["observer", "excursion1"]),
            "observer_exc2": opGet(["observer", "excursion2"]),
            "observer_exc3": opGet(["observer", "excursion3"]),
            "observer_exc4": opGet(["observer", "excursion4"]),

            # "teams": [ teamIndex(team) for team in opGet(["teams"])],
            "teams": opGet(["teams"]),
            **getUserInfo(opGet(["userinfo"])),
        })
    return result

def getTeam(documents): 
    i = 0
    result = []
    for document in documents :
        opGet = lambda args: optionalGet(document, args) 
        if i%100 == 0: print(i, 'rows done')
        i += 1
        result.append({
            "id": opGet(["_id"]),
            "team_submit": opGet(["info", "submit"]),

            "team_school": opGet(["info", "school"]),
            "team_address": opGet(["info", "address"]),
            "team_contactname": opGet(["info", "contactname"]),
            "team_contactemail": opGet(["info", "contactemail"]),
            "team_checkin": opGet(["info", "checkin"]),
            "team_room": opGet(["info", "room"]),
            "team_excursion1": opGet(["info", "excursion1"]),
            "team_excursion2": opGet(["info", "excursion2"]),
            "team_excursion3": opGet(["info", "excursion3"]),
            "team_excursion4": opGet(["info", "excursion4"]), 

            **{'c1_'+k:v for (k, v) in getUserInfo(opGet(["contestant1"])).items()}, 
            **{'c2_'+k:v for (k, v) in getUserInfo(opGet(["contestant2"])).items()}, 
            **{'c3_'+k:v for (k, v) in getUserInfo(opGet(["contestant3"])).items()}, 
        })
    return result

load_dotenv('../.env')


conn_str = os.environ.get("MONGOURI")
if not conn_str : 
    print("no database config in the environment")
    exit()
client = pymongo.MongoClient(conn_str, serverSelectionTimeoutMS=5000)

db = client["test"]
collection = db["users"]
documents = collection.find()

users = getUser(documents)
print('writing')

FIELDSNAME = users[0].keys()
with open('user.csv', 'w+', newline='') as f:
    writer = csv.writer(f)
    writer.writerow(FIELDSNAME)

    for user in users :
        writer = csv.DictWriter(f, fieldnames=FIELDSNAME)
        writer.writerow(user)


db = client["test"]
collection = db["teams"]
documents = collection.find()

teams = getTeam(documents)

print('writing')

FIELDSNAME = teams[0].keys()
with open('team.csv', 'w+', newline='') as f:
    writer = csv.writer(f)
    writer.writerow(FIELDSNAME)

    for team in teams :
        writer = csv.DictWriter(f, fieldnames=FIELDSNAME)
        writer.writerow(team)
